File: src/main/java/me/ste/stevesseries/fancydrops/packet/PacketPlayOutSpawnEntity.java
package me.ste.stevesseries.fancydrops.packet;

import com.comphenix.protocol.PacketType;
import com.comphenix.protocol.ProtocolLibrary;
import com.comphenix.protocol.events.PacketContainer;
import me.ste.stevesseries.fancydrops.item.FancyItemArmorStand;
import org.bukkit.Location;
import org.bukkit.entity.EntityType;
import org.bukkit.util.Vector;

import java.util.UUID;

public class PacketPlayOutSpawnEntity {
    private final int entityId;
    private final UUID entityUuid;
    private final EntityType entityType;
    private final Location location;
    private final int data;
    private final Vector velocity;

    public PacketPlayOutSpawnEntity(int entityId, UUID entityUuid, EntityType entityType, Location location, int data, Vector velocity) {
        this.entityId = entityId;
        this.entityUuid = entityUuid;
        this.entityType = entityType;
        this.location = location;
        this.data = data;
        this.velocity = velocity;
    }

    public PacketPlayOutSpawnEntity(FancyItemArmorStand stand) {
        this(stand.getEntityId(), stand.getEntityUuid(), EntityType.ARMOR_STAND, stand.getLocation(), 0, new Vector(0, 0, 0));
    }

    public PacketContainer toContainer() {
        PacketContainer packet = ProtocolLibrary.getProtocolManager().createPacket(PacketType.Play.Server.SPAWN_ENTITY);
        packet.getIntegers().write(0, this.entityId); // Entity ID
        packet.getUUIDs().write(0, this.entityUuid); // Entity UUID
        packet.getEntityTypeModifier().write(0, this.entityType); // Entity Type
        packet.getDoubles().write(0, this.location.getX()); // Position X
        packet.getDoubles().write(1, this.location.getY()); // Position Y
        packet.getDoubles().write(2, this.location.getZ()); // Position Z
        packet.getIntegers().write(4, (int) (this.location.getYaw() * 256.0F / 360.0F)); // Yaw
        packet.getIntegers().write(5, (int) (this.location.getPitch() * 256.0F / 360.0F)); // Pitch
        packet.getIntegers().write(6, this.data); // Data
        packet.getIntegers().write(1, (int) (this.velocity.getX() * 8000)); // Velocity X
        packet.getIntegers().write(2, (int) (this.velocity.getY() * 8000)); // Velocity Y
        packet.getIntegers().write(3, (int) (this.velocity.getZ() * 8000)); // Velocity Z
        return packet;
    }
}

File: src/main/kotlin/me/ste/stevesseries/fancydrops/packet/PacketPlayOutSpawnEntity.kt
package me.ste.stevesseries.fancydrops.packet

import com.comphenix.protocol.PacketType
import com.comphenix.protocol.ProtocolLibrary
import com.comphenix.protocol.events.PacketContainer
import net.minecraft.util.MathHelper
import org.bukkit.Location
import org.bukkit.entity.EntityType
import org.bukkit.util.Vector
import java.lang.Double.max
import java.lang.Double.min
import java.util.*

class PacketPlayOutSpawnEntity(
    private val entityId: Int,
    private val entityUuid: UUID,
    private val entityType: EntityType,
    private val location: Location,
    private val data: Int,
    private val velocity: Vector
) : Packet() {
    override val container: PacketContainer
        get() {
            val packet = ProtocolLibrary.getProtocolManager().createPacket(PacketType.Play.Server.SPAWN_ENTITY)

            packet.integers.write(0, this.entityId)
            packet.uuiDs.write(0, this.entityUuid)
            packet.entityTypeModifier.write(0, this.entityType)
            packet.doubles.write(0, this.location.x)
            packet.doubles.write(1, this.location.y)
            packet.doubles.write(2, this.location.z)
            packet.integers.write(1, (max(-3.9, min(3.9, this.velocity.x)) * 8000.0).toInt())
            packet.integers.write(2, (max(-3.9, min(3.9, this.velocity.y)) * 8000).toInt())
            packet.integers.write(3, (max(-3.9, min(3.9, this.velocity.z)) * 8000).toInt())
            packet.bytes.write(0, (this.location.pitch * 256F / 360F).toInt().toByte())
            packet.bytes.write(1, (this.location.yaw * 256F / 360F).toInt().toByte())
            packet.bytes.write(2, (this.location.yaw * 256F / 360F).toInt().toByte()) //headYaw
            packet.integers.write(4, this.data)

            return packet
        }
}